src/Raven.Server/Documents/ETL/Providers/AI/Embeddings/EmbeddingsGenerationTask.cs

```diff
@@ -0,0 +1,487 @@
+                            _missingEmbeddingsHolder.Add(embeddingToGenerate.TextualValue, embeddingToGenerate);
+                        else
+                            embeddingsInCache++;
+                    }
+                }
+
+                processed++;
+            }
+
+            var embeddingsMap = _missingEmbeddingsHolder.GetEmbeddingsMap();
+            if (embeddingsMap.Keys.Any())
+            {
+                var keys = embeddingsMap.Keys.ToList();
+
+                if (Database.AiIntegrations.TryGetServiceByConnectionString(connectionStringId, out var service) == false)
+                    throw new ArgumentException($"Couldn't find embedding generation service for connection string '{connectionStringId.Value}'");
+
+                IList<ReadOnlyMemory<float>> generatedValues;
+                try
+                {
+                    using var embeddingsGenerationScope = scope.For(EmbeddingsGenerationOperations.GenerateInAiService);
+                    generatedValues = AiHelper.GenerateEmbeddingsAsync(service.Instance, keys).GetAwaiter().GetResult();
+                    embeddingsGenerationScope.NumberOfGeneratedEmbeddings = generatedValues.Count;
+                    embeddingsGenerationScope.NumberOfEmbeddingsInCache = embeddingsInCache;
+
+                }
+                catch (Exception ex)
+                {
+                    if (ex is HttpOperationException { StatusCode: HttpStatusCode.TooManyRequests })
+                    {
+                        throw new EmbeddingGenerationException(
+                            $"Failed to generate embeddings due to rate limits. " +
+                            $"The process will increase the delay between calls to the model using {Database.Configuration.Ai.EmbeddingsGenerationTaskRetryStrategy.ToString().ToLower()} backoff strategy. " +
+                            $"However, decreasing the number of elements processed in a single batch (using '{RavenConfiguration.GetKey(x => x.Ai.EmbeddingsGenerationTaskMaxBatchSize)}' " +
+                            $"configuration option) may help, or you can increase the limits on your model deployment.",
+                            ex);
+                    }
+
+                    throw;
+                }
+
+                if (generatedValues.Count != keys.Count)
+                    throw new InvalidOperationException($"Generated embeddings count ({generatedValues.Count}) does not match missing values count ({keys.Count})");
+
+                for (var i = 0; i < keys.Count; ++i)
+                {
+                    var key = keys[i];
+                    var embedding = EmbeddingsHelper.CreateEmbeddingValue(generatedValues[i], Configuration.Quantization);
+
+                    foreach (var embeddingItem in embeddingsMap[key])
+                    {
+                        embeddingItem.SetEmbedding(embedding, Configuration.Quantization, connectionStringId);
```

Comment:
I don't follow why we have `SetEmbedding` here and we need to keep track of each one of them